### ISSUE TITLE
EPIC: patternfly.css performance issues

### ISSUE THREAD

**Issue by mattnolting** [patternfly/patternfly#6694]
# TLDR (from #6369)
patternfly.css has performance issues. These issues are due to global selectors `*`, `:has()`, `:is()`, `:where()`, `:first/last-child`, references to elements rather than selectors, and complex selector chains. Here is our current [performance profile](https://docs.google.com/spreadsheets/d/1NuHHMuCDu4mmxmRkj-MJYqDFXbOyv5phddtBZAaHdOQ/edit#gid=0). It's clear that these issues need to be addressed to optimize our css and through several updates to our codebase, we can do just that. 

## Update: 8/19/24

As I was recorded performance metrics, I found interesting results. While attempting to compare metrics (`match attempts/count` increase/decrease).

![Image](https://github.com/user-attachments/assets/683c9db8-59d9-4d61-a93e-8c204cf70a6f)

I realized that that the selectors I was looking for were no longer available. 

![Image](https://github.com/user-attachments/assets/5f2e6b6e-8e2e-4702-9852-d4b58d8f65fa)

The comparison I was looking for didn't exist because it was completely eliminated. 

By removing `:where(:root, .component)` alone,  `match attempts` went from 6,516,114 to 541,945, resulting in a total reduction of 5,974,169 `match attempts`. 

## Update: 7/30/24

There appear to be two, core causes for performance issues. As noted in @garrett's [issue](https://github.com/patternfly/patternfly/issues/6369), our approach to floating pseudo elements, global selectors, html elements, etc has a significant impact upon First Contentful Paint, Largest Contentful Paint, and Speed Index. The second, devtools based issue is part of our workspace JS.  

## Related content
* https://blogs.windows.com/msedgedev/2023/01/17/the-truth-about-css-selector-performance/
* https://stackoverflow.com/questions/5797014/why-do-browsers-match-css-selectors-from-right-to-left/5813672#5813672
* https://www.stevesouders.com/blog/2009/06/18/simplifying-css-selectors/
* https://webperf.tips/tip/layout-thrashing/

### Pseudo selectors
* `:has()` is bad, don't use it ever unless applied directly to class or id
* floating `:where()` pseudos are gobbling up memory. `:where(:root)` tells browser engines to look at every element on the page. Meaning it's the same as `*` selector. 
* `:is()` adds quite a lot of page load time and should be avoided
* remove `:first/last-child` psuedos

### Complex selectors
* complex selectors like `.pf-v6-c-table tbody:where(.pf-v6-c-table__tbody) > tr:where(.pf-v6-c-table__tr) > :where(th, td)` need to be updated. 
* `.pf-v6-c-table__tbody > .pf-v6-c-table__tr > .pf-v6-c-table__th, .pf-v6-c-table__tbody > .pf-v6-c-table__tr > .pf-v6-c-table__td` is much more efficient for browsers engines

### Global selectors
* update flex and grid, remove `> *` this will like mean that `pf-v6-l-flex` will require `column/row-gap` updates 
* update boilerplate

### Breakpoints
* breakpoint loops add significant bloat to `patternfly.css`. This is where `resize-observer` would be a critical addition to our codebase as it runs once after element resize and updates individual selectors vs browser engine looking for styles at `-sm, -md-, etc`

* avoid [forced reflow](https://webperf.tips/tip/layout-thrashing/) 

### `@at-root` 
• `@at-root` compiles to the root stylesheet, rather than the namespaced stylesheet. This makes `@at-root` rules present on every screen, rather than the contextual screen. 

## Performance updates:

Issues are currently on hold and available to reference until we have a plan to implement

- [x] #6722
- [x] #7032 
- [ ] #7146
- [ ] #7033
- [ ] #6721
- [ ] #6720
- [ ] #6723
- [ ] #6724
- [ ] #6725
- [ ] #6927
- [ ] #6942
- [x] #6941

## Key takeaways **JS performance**:

[Long tasks](https://web.dev/articles/optimize-long-tasks?utm_source=devtools)

## Key takeaways **Font performance**:

Use WOFF2 format: It is the most optimized and widely supported format.
Preload key fonts: Use <link rel="preload" href="path/to/font.woff2" as="font" type="font/woff2" crossorigin="anonymous"> to improve loading performance.

**Comment by srambach** [comment]
@mattnolting Thanks for making this list. I'm going to mark these issues as "needs info" and we should make sure each one has:
- why the change is necessary (what specific problem it's addressing)
- what the expected improvement should be (metrics if possible)
- risks and mitigation
- how we prove that it hasn't introduced any problems (screen shot testing, or ?)
Once that's there, as a team we should review and prioritize.
And then in the PRs let's make sure to document metrics before and after the change.

**Comment by mcoker** [comment]
Closing this issue for now and leaving the parent issue open in parking lot in case we want to address again. We can review the open PRs against this issue to see what impact the changes have and if we want to make any changes, we can either open new PRs or try to update the existing PR if it isn't too far from main.
